tests/test_raft.rs

```diff
@@ -744,6 +1005,58 @@ fn test_dueling_candidates() {
 }
 
 #[test]
+fn test_dueling_pre_candidates() {
+    let mut a = new_test_raft(1, vec![1, 2, 3], 10, 1, new_storage());
+    let mut b = new_test_raft(2, vec![1, 2, 3], 10, 1, new_storage());
+    let mut c = new_test_raft(3, vec![1, 2, 3], 10, 1, new_storage());
+    a.pre_vote = true;
+    b.pre_vote = true;
+    c.pre_vote = true;
+
+    let mut nt = Network::new_with_config(vec![Some(a), Some(b), Some(c)], true);
+    nt.cut(1, 3);
+
+    nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+
+    // 1 becomes leader since it receives votes from 1 and 2
+    assert_eq!(nt.peers[&1].state, StateRole::Leader);
+
+    // 3 campaigns then reverts to follower when its pre_vote is rejected
+    assert_eq!(nt.peers[&3].state, StateRole::Follower);
+
+    nt.recover();
+
+    // Candidate 3 now increases its term and tries to vote again.
+    // With pre-vote, it does not disrupt the leader.
+    nt.send(vec![new_message(3, 3, MessageType::MsgHup, 0)]);
+    let wlog = new_raft_log(vec![empty_entry(1, 1)], 2, 1);
+    let wlog2 = new_raft_log_with_storage(new_storage());
+    let tests = vec![
+        (StateRole::Leader, 1, &wlog),
+        (StateRole::Follower, 1, &wlog),
+        (StateRole::Follower, 1, &wlog2),
+    ];
+    for (i, &(state, term, raft_log)) in tests.iter().enumerate() {
+        let id = i as u64 + 1;
+        if nt.peers[&id].state != state {
+            panic!("#{}: state = {:?}, want {:?}",
+                   i,
+                   nt.peers[&id].state,
+                   state);
+        }
+        if nt.peers[&id].term != term {
+            panic!("#{}: term = {}, want {}", i, nt.peers[&id].term, term);
+        }
+        let base = ltoa(raft_log);
+        let l = ltoa(&nt.peers[&(1 + i as u64)].raft_log);
+        if base != l {
```

Comment:
This piece of code is to ensure two logs have the same applied index, commit index,  and log entries. In TiKV, it simply use `String::ne` instead of `diffu` in every test case.